witness/src/Dscp/Snowdrop/Storage/Avlp.hs

```diff
@@ -0,0 +1,311 @@
+
+initAVLStorage
+    :: forall h k v m .
+    ( AVLMutates' h k v m )
+    => Map k v -> m ()
+initAVLStorage (Map.toList -> kvs) = do
+    AVL.initialiseStorageIfNotAlready @h kvs
+
+-- | DbModifyActions with given plugin - rdb one or pure.
+mkAvlDbModifyActions
+  :: forall h k v t m n
+  .  ( MonadIO n
+     , IdSumPrefixed k
+     , AVLMutates h k v t m
+     )
+  => DBPlugin t
+  -> n (DbModifyActions (AVLChgAccum h k v) k v m (AVL.Proof h k v))
+mkAvlDbModifyActions plugin@ (DBPlugin run) = do
+    lookups <- newTVarIO (Set.empty :: Set AVL.Revision)
+    return DbModifyActions
+        { dmaAccessActions = mkAvlDbAccessActions plugin lookups
+        , dmaApply         = mkDmaApply                   lookups
+        }
+  where
+    -- Apply changes to the database.
+    mkDmaApply lookups accum' = do
+      accum        <- initAccum plugin accum'
+      nodesTouched <- atomically $ STM.swapTVar lookups Set.empty
+      old          <- run AVL.currentRoot
+      let preproof =  nodesTouched <> acaTouched accum
+      proof        <- run $ AVL.prune preproof (AVL.fullRehash old)
+
+      run $ AVL.append (acaMap accum)
+      return proof
+
+-- | Readonly interface for given plugin.
+mkAvlDbAccessActions
+  :: forall h k v t m
+  .  ( IdSumPrefixed k
+     , AVLMutates h k v t m
+     )
+  => DBPlugin t
+  -> TVar (Set.Set AVL.Revision)
+  -> DbAccessActions (AVLChgAccum h k v) k v m
+mkAvlDbAccessActions plugin@ (DBPlugin run) lookups = DbAccessActions
+    { daaGetter
+    , daaModifyAccum
+    , daaIter
+    }
+  where
+    storeNodeset :: Set.Set AVL.Revision -> m ()
+    storeNodeset nodeset = do
+        atomically $ STM.modifyTVar lookups (<> nodeset)
+
+    daaGetter :: AVLChgAccum h k v -> StateR k -> m (StateP k v)
+    daaGetter accum' req = do
+        accum <- initAccum plugin accum'
+        let
+          getOne newAccum key = do
+            ((val, proof), _) <- run $ AVL.lookup key (acaMap newAccum)
+            storeNodeset proof
+            return $ maybe Map.empty (Map.singleton key) val
+
+        ss <- mapM (getOne accum) (Set.toList req)
+        return $ Map.unions ss
+
+    daaIter
+        :: forall b
+        .  AVLChgAccum h k v
+        -> Prefix
+        -> b
+        -> ((k, v) -> b -> b)
+        -> m b
+    daaIter accum' prefix start folder = do
+        accum <- initAccum plugin accum'
+        (result, nodeset) <- run $ AVL.foldIf (\x -> idSumPrefix x == prefix, start, folder, id) (acaMap accum)
+        storeNodeset nodeset
+        return result
+
+    daaModifyAccum
+        :: AVLChgAccum h k v
+        -> ChgAccumModifier k v
+        -> m (Either (CSMappendException k) (AVLChgAccum h k v, Undo k v))
+    daaModifyAccum oldAccum changes = do
+        let cSet = case changes of
+              CAMChange       cs       -> cs
+              CAMRevert (Undo cs _wat) -> cs
+
+        applyChangeSet plugin cSet oldAccum
+
+-- | Apply changes to the accum, using the storage plugin.
+apply
```

Comment:
I don't mean to be boring, but it's better to use less general naming so the jump-to-definition doesn't go crazy.